### ISSUE TITLE
Hibernate ORM extension should save reflection to create/access fields of entities

### ISSUE THREAD

**Issue by franz1981** [quarkusio/quarkus#43692]
### Description

Currently https://github.com/quarkusio/quarkus/blob/ab8f230bffd54b9e76e119e54f3bca40718be152/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/service/bytecodeprovider/RuntimeBytecodeProvider.java#L33 suggests that Hibernate is using reflection-based accessors to get/set ORM entities fields: just running https://github.com/quarkusio/quarkus-quickstarts/tree/242213b09ca8cca7cb4f3a6dda783516f104b14a/hibernate-orm-quickstart confirm this theory (I'll attach some flamegraphs from the infamous Techempower's fortune below, to show it).

Similarly, constructing entities is performed via `org/hibernate/metamodel/internal/EntityInstantiatorPojoStandard.instantiate`,  which again relies on reflection to perform the entities allocations.

Ideally both field accesses and allocation(s) should be performed at build-time: for the former it means implementing build-time `AccessOptimizer`, similarly to what `Qute` does for its resolver(s) and providing it via a `ReflectionOptimizer::getAccessOptimizer`.

While, for the allocation-side, it is performed by `org/hibernate/metamodel/internal/EntityInstantiatorPojoStandard.instantiate` which "maybe" (I need some hibernate dev to confirm this) need `getRepresentationStrategy().getInstantiator()` to return a custom build-time generated implementation which call the right constuctor.

**Comment by mbladel** [comment]
> which "maybe" (I need some hibernate dev to confirm this) need getRepresentationStrategy().getInstantiator() to return a custom build-time generated implementation

Hey @franz1981, it is indeed possible to use a custom `EntityInstantiator`, which is specified by the `ReflectionOptimizer::getInstantiationOptimizer` method:

https://github.com/hibernate/hibernate-orm/blob/f68f6aae50b0f2fb371cceb65cd8812af76b734f/hibernate-core/src/main/java/org/hibernate/metamodel/internal/EntityRepresentationStrategyPojoStandard.java#L197-L208

We also have a default bytecode-enhancement reflection optimizer [implementation](https://github.com/hibernate/hibernate-orm/blob/a92bf606a9d29ee061ed37bd5f88ead41ea76631/hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/BytecodeProviderImpl.java#L136) within hibernate, but apparently the Qarkus extension is explicitly avoiding that.

**Comment by franz1981** [comment]
Thanks @mbladel for confirming it!

**Comment by mbladel** [comment]
Hello @franz1981, so I did some initial experimentation with what is available right now, that being the `org.hibernate.bytecode.spi.ReflectionOptimizer` contract within Hibernate ORM, hoping we could create the required `ReflectionOptimizer.InstantiationOptimizer` and `ReflectionOptimizer.AccessOptimizer` bytecode at build time and just pass those through to use at runtime.

While the instantiation optimizer would be very easy to create, simply relying on finding a no-arg constructor in every mapped entity class and creating a utility that calls it, the same is not true for property access.

Sadly, there are some pre-requisites needed to build access-optimizers: Hibernate currently expects them to retrieve _all_ properties from a mapped entity class, where "properties" might not simply correspond to the values of the attributes contained in a class, read through fields or getters, but also some special values handled by Hibernate itself at runtime. 
These "synthetic properties" are generated through the building of `Metadata` at Hibernate's boot time, which as we know is not going to be available at Quarkus' build-time as of now (though there is [some work](https://github.com/quarkusio/quarkus/issues/40897) going on with Hibernate 7.0 that might hopefully change that in the future).

Within my experimentation I was able to build access optimizers at static-init, i.e. after we build Hibernate's `Metadata`, but this would have an impact on startup performance. We already do some very limited bytecode generation there for Hibernate proxies that were not generated at boot, with plenty of warnings, and I'm not sure if it's a good idea to add more. If reflection has that big of an impact, though, this might be an option for the time being (note: this will require some work on the Hibernate side to avoid duplicating logic).

@yrodiere and I also discussed this, and the first thing he suggested would be moving Hibernate to use method handles instead of reflection. From what we understand, this would already prevent many aspects of the performance impact that reflection has, though we are not sure if it would be enough. Do you think that moving to method handles (even with the optimization of generating them at build-time to be referenced within  `static final` fields) would be enough to justify not having bytecode-enhanced access optimization?

**Comment by franz1981** [comment]
> Do you think that moving to method handles (even with the optimization of generating them at build-time to be referenced within static final fields) would be enough to justify not having bytecode-enhanced access optimization?

If they are going to be generated in static final, yep, this should help. But the problem in that case will be the morphism i.e. each property access represent a different concrete type of method handle which will still cause if we have more than 2 types, a megamorphic virtual call, which is still not ideal. In short, it is an improvement, but we need to measure if it is enough to help

**Comment by mbladel** [comment]
Right, we also need  to understand the possible impacts and requirements for migrating Hibernate ORM to method handles in the first place, but that's something we can explore for the future.

About the other "option", i.e. creating the access optimizers at static-init, do you think the additional work at startup would be justified by the gains in runtime performance by avoiding reflection @franz1981? Or is that something not even worth exploring since it would penalize boot times too much?

**Comment by franz1981** [comment]
> Or is that something not even worth exploring since it would penalize boot times too much?

There is another aspect related feature parity requirement which won't make a runtime bytecode generation so appealing and it is related native image behaviour here; I have noticed that we tried hard to keep native image and JIT to have similar "behaviour" - and this is especially true if, by implementing this optimization, we risk to go outside what is handled in native image as well (reflection, string concatenation, record equals/hashCode, ...). 
So, my current feeling is that such difference is not what we want people to have, but I let @cescoffier and @geoand to give their view.
Atm getting something at build time and no impacts at runtime is what we really want to seek, imo.

**Comment by franz1981** [comment]
@mbladel @yrodiere 
after speaking with @Sanne It looks like reflection shouldn't happen - and it wasn't happening, in some past point in time, in Quarkus: so it requires some further investigation and this can be seen as a bug, not an improvement - since it was working, before.

**Comment by yrodiere** [comment]
@Sanne can you please clarify what piece of code was ( in the past) supposed to avoid reflection in Hibernate ORM? As far as I can see the reflection optimizer was never used in Quarkus, and I'm not aware of any other hack/feature that would achieve the same effect.

@franz1981 can you confirm that running your reproducer using an older version of Quarkus/Hibernate will make the reflection use disappear?

**Comment by Sanne** [comment]
I'm sorry I don't remember on top of my head - I *think* we had worked on this in the early days as we had a stronger opinion about attempting to avoid any reflection at all, a stance which got a bit more relaxed and pragmatic as we grew.
And I don't remember seeing the reflective calls when we traced them - but it's been quite some time and I might be wrong.

I'll try to dig it up.

I do remember that we never got to optimize the instantiators - that is still on my wishlist as well.

**Comment by mbladel** [comment]
@Sanne as Yoann already said, I don't see how reflection would've been avoided in Hibernate since the `ReflectionOptimizers` were never returned by the runtime bytecode provider, though I don't have a memory of how it worked before Hibernate 6 so I could be missing something.

Me and @franz1981 worked on a [prototype](https://github.com/quarkusio/quarkus/pull/43767) creating instantiation optimizers, but benchmarks showed that they might not even be worth it: when more then a couple entities are involved, invoking optimizers result in megamorphic virtual calls that basically nullify the benefits gained from avoiding reflection (and with the type pollution bugfix, reflection ends up performing even slightly better), see also the [benchmarks pr](https://github.com/hibernate/hibernate-orm-benchmark/pull/13#issuecomment-2441217589) for more details.

**Comment by Sanne** [comment]
Right, I'm aware of the challenges with instantiation optimizers; they might need a more thourough rethink in ORM.

Regarding the `ReflectionOptimizers`, the one in Quarkus returning null today is not the one which was intended to be used.

**Comment by yrodiere** [comment]
> I'm sorry I don't remember on top of my head - I _think_ we had worked on this in the early days as we had a stronger opinion about attempting to avoid any reflection at all, a stance which got a bit more relaxed and pragmatic as we grew. And I don't remember seeing the reflective calls when we traced them - but it's been quite some time and I might be wrong.
> 
> I'll try to dig it up.

I stumbled upon this while working on something else:

https://github.com/quarkusio/quarkus/blob/8fd4337ee1160ace2aec3c1f4f61377a3db96213/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaEntitiesBuildItem.java#L39-L43

It would seem we've always registered entity fields/methods for reflection, as far back as 2019. So I doubt we've ever had anything to avoid reflection.

**Comment by franz1981** [comment]
@mbladel differently from the constructor benchmark at https://github.com/hibernate/hibernate-orm-benchmark/commit/f1431eb9c1d26af82a3edb727ac5ed63441fb713 if we can build one using fields will help to understand the benefits for such (which I can expect to be relevant - TBH - since it's more "batchy" and virtual dispatching shouldn't absorb the most of the improvement), such could help driving prioritization for this "new" feature - which will fill the gap between what Quarkus deliver vs Wildfly (latter one is doing it as expect - reflection-free).
Additionally this should benefit native-mode as well, but we need a specific benchmark for such (@galderz  :P )

**Comment by geoand** [comment]
> like people exposing entities over Jackson

Please allow me to go on a minor  tangent: I hope that at some point we either come up with a solution for this (I guess a long the line that @yrodiere proposed elsewhere) or provide some clear guidance to users on how to deal with such scenarios.

**Comment by mbladel** [comment]
> [@mbladel](https://github.com/mbladel) differently from the constructor benchmark at [hibernate/hibernate-orm-benchmark@f1431eb](https://github.com/hibernate/hibernate-orm-benchmark/commit/f1431eb9c1d26af82a3edb727ac5ed63441fb713) if we can build one using fields will help to understand the benefits for such

Hey @franz1981, I finally got around to looking into this (sorry for the delay, holidays and other priorities took over). You can find a first draft of a benchmark which compares reflective field access vs bytecode optimizations here: https://github.com/hibernate/hibernate-orm-benchmark/pull/14.

I've described my findings in detail in the PR description, but the TL;DR is we see a significant improvement of about 15% on a group of 4 small to moderately sized entities (5 - 30 properties) when queries the database and filling new entity instances with values. Performance on a single entity type with very few properties is comparable to very slightly better.

**Comment by franz1981** [comment]
@mbellade this is still opened? any changes which we've missed?

**Comment by mbellade** [comment]
It is @franz1981, sadly at the moment we are unable to build the necessary access optimizers to avoid reflection when accessing entity class fields. We're waiting on some progress with https://github.com/quarkusio/quarkus/issues/40897, since Hibernate 7 we should be able to move a lot more of Hibernate's startup processes to either build or static-init time, and hopefully we might have enough information to build these optimizers - though it's too early to say.

**Comment by franz1981** [comment]
A very important update @Sanne @marko-bekhta @mbellade 

While investigating why Quarkus virtual threads don't outperform platform threads on a JDBC/Hibernate workload (GET /fruits, 10 fruits + 32 store prices, 4 cores, 100 connections), syncjfr profiling shows `java.lang.invoke.Invokers.checkCustomized` as **6% of CPU samples** in Q3-Virtual but **0% in Q3-JVM**.

The samples come from Hibernate's entity hydration path: `Field.set` → `MethodHandleObjectFieldAccessorImpl.set` → `invokeExact_MT` → `checkCustomized` (14+7 samples) and `Constructor.newInstance` → `DirectConstructorHandleAccessor.newInstance` → `invokeExact_MT` → `checkCustomized` (5 samples).

Since JDK 18 (JEP 416), core reflection (`Field.set`, `Constructor.newInstance`) is implemented on top of MethodHandles. The JIT eliminates `checkCustomized` when the MethodHandle is a compile-time constant (`isCompileConstant` intrinsic). With platform threads, the JIT succeeds — `checkCustomized` never appears as a leaf frame. With virtual threads, it doesn't — likely because VTs migrating between ForkJoinPool carriers disrupt the JIT's profile data, preventing constant-folding.

**Potential fix**: ensure Hibernate entity field access goes through Quarkus's build-time generated `HibernateAccessOptimizer` classes (direct field access, no `Field.set()`) for all fields and constructors. This bypasses the JEP 416 MethodHandle reflection layer entirely. If reflection is unavoidable, caching the resolved `MethodHandle` in a `static final` field would let the JIT constant-fold it regardless of VT carrier migration.

Related: Jetty [#6328](https://github.com/eclipse/jetty.project/issues/6328) hit the same `checkCustomized` overhead (30% CPU) from per-connection `MethodHandle.bindTo()`.

**Comment by geoand** [comment]
> With virtual threads, it doesn't — likely because VTs migrating between ForkJoinPool carriers disrupt the JIT's profile data, preventing constant-folding.

Oh man, who would have thought that JVM platform features interact in so many different ways...

Great find @franz1981!

**Comment by FroMage** [comment]
So, you're saying that switching from reflection to using `MethodHandle` to set the field would not be enough, we'd have to generate bytecode to access the fields?

**Comment by yrodiere** [comment]
> So, you're saying that switching from reflection to using `MethodHandle` to set the field would not be enough, we'd have to generate bytecode to access the fields?

Good thing that's exactly what @marko-bekhta, @dreab8, @mbellade are working on, then.

**Comment by FroMage** [comment]
I am surprised though, because I remember being told that we could use `MethodHandle` instead of reflection and bytecode generation, that it was equivalent in performance to bytecode, and saved on code size/class count.

**Comment by FroMage** [comment]
Oh, hold on, so if the `MethodHandle` is in a `static final` field then the perf is equivalent? OK. Well, I feel like that's a bit of important lore right there, that we should write down somewhere.

**Comment by franz1981** [comment]
https://timefold.ai/blog/java-reflection-but-much-faster is a very nice old wrap up on this ^^
Consider that now reflection is built on to of MH so that specific part is no longer true

**Comment by FroMage** [comment]
So we pay at startup the cost of loading all these generated accessor classes, that's what the article points out. Looks to me like it's still a better option to 🤦 generate a class file with `max field` static fields pointing to the `MethodHandle`, then, no?

**Comment by mbellade** [comment]
If we're worried about the impact of loading lots of classes, but we're generating bytecode anyway, we could also generate accessor bridges with `static` methods that don't use `MethodHandle`s at all but directly access fields/methods/constructors for the target classes. That would avoid the cost entirely and we don't need to depend on JIT at all.

**Comment by FroMage** [comment]
But how do you invoke those static methods?

**Comment by mbellade** [comment]
Either through custom Hibernate `AccessOptimizer`s or `org.hibernate.property.access.spi.Setter/Getter`s that know they can rely on them. These would be created at boot time or, ideally, even at static-init (but for the latter, again, we'll need to move more of the Hibernate bootstrap metadata earlier in the Quarkus build).

**Comment by FroMage** [comment]
If they implement one of those types, they're a class, no? I must be missing something.

**Comment by mbellade** [comment]
Of course, but it's one class per entity type, instead of one per property. That's just unavoidable, for how Hibernate ORM's property access SPIs work at the moment.

**Comment by FroMage** [comment]
Oh OK, yeah that's fine.
